src/imputation/apportionment.py

```diff
@@ -0,0 +1,173 @@
+"""Implement apportionment of headcount and FTE."""
+import logging
+import pandas as pd
+import numpy as np
+
+from typing import Dict, List
+
+ApportionmentLogger = logging.getLogger(__name__)
+
+
+def calc_202_totals(df: pd.DataFrame) -> pd.DataFrame:
+    """Calculate totals of q202 by reference.
+
+    For each reference, the sum over all instances for q202 is 
+    calculated and entered in column qtot_202_all for each instance. 
+    Also, the sums over all instances for just the civil or just 
+    defence is entered in column qtot_202_CD.
+    
+    Args:
+        df (pd.DataFrame): The main dataset for apportionment.
+
+    Returns:
+        pd.DataFrame: The main dataset with totals.
+    """
+    df["tot_202_all"] = df.groupby("reference")["202"].transform(sum)
+
+    df["tot_202_CD"] = df.groupby(["reference", "200"])["202"].transform(sum)
+
+    return df
+
+
+def update_column(df: pd.DataFrame, col: str) -> pd.Series:
+    """Copy item in insance 0 to all other instances in a given reference.
+
+    For long form entries, questions 405 - 412 and 501 - 508 are recorded
+    in instance 0. A series is returned representing an updated version of the 
+    column with the values from instance 0 copied to all other instances.
+
+    Args:
+        df (pd.DataFrame): The main dataset for apportionment.
+        col (str): The name of the 4xx or 5xx column being treated.
+
+    Returns:
+        pd.Series: A series with the values in instance 0 copied to
+            other instances for the same reference.
+    """
+    updated_col = df.groupby("reference")[col].transform(sum)
+    return updated_col
+
+
+def calc_fte_column(
+    df: pd.DataFrame, 
+    fte_dict: Dict[str, List[str]], 
+    round_val=4
+) -> pd.DataFrame:
+    """Create a new column for FTE values.
+
+    Args:
+        df (pd.DataFrame): The main dataset for apportionment.
+    Returns:
+        pd.Dataframe: The dataset with one new column for FTE.    
+    """
+
+    # create new apportionment column for the civil cases (avoid division by 0)
+    for new_col, old_cols in fte_dict.items():
```

Comment:
One suggestion to avoid repeated code:
```suggestion
    categories = [["C", 0], ["D", 1]]

    for new_col, old_cols in fte_dict.items():
        for category in categories:
            filter_condition = (df["200"] == category[0]) & (df["tot_202_CD"] > 0)
            df.loc[filter_condition, new_col] = round(
                update_column(df, old_cols[category[1]]) * df["202"] / df["tot_202_CD"],
                round_val
            )
    return df
```

Reply:
Good suggestion, I have made the change.